lib/services/local/resource.go

```diff
@@ -0,0 +1,365 @@
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package local
+
+import (
+	"context"
+	"github.com/gravitational/teleport/lib/backend"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/trace"
+)
+
+// CreateResources attempts to dynamically create the supplied resources.
+// This function returns `trace.AlreadyExistsError` if one or more resources
+// would be overwritten, and `trace.NotImplementedError` if any resources
+// are of an unsupported type (see `ItemizeResource(...)`).
+//
+// NOTE: This function is non-atomic and performs no internal synchronization;
+// backend must be locked by caller when operating in parallel environment.
+func CreateResources(b backend.Backend, resources ...services.Resource) error {
+	var items []*backend.Item
+	// itemize all resources & ensure that they do not exist.
+	for _, r := range resources {
+		item, err := ItemizeResource(r)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		_, err = b.Get(context.TODO(), item.Key)
+		if !trace.IsNotFound(err) {
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			return trace.AlreadyExists("resource %q already exists", string(item.Key))
+		}
+		items = append(items, item)
+	}
+	// create all items.
+	for _, item := range items {
+		_, err := b.Create(context.TODO(), *item)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
+
+// ItemizeResource attempts to construct an instance of `backend.Item` from
+// a given resource.  If `rsc` is not one of the supported resource types,
+// a `trace.NotImplementedError` is returned.
+func ItemizeResource(resource services.Resource) (*backend.Item, error) {
```

Comment:
I would call it `ResourceToItem` instead of `Itemize`, both are correct, but I think we use the first notation more frequently throughout the codebase, similarly, `ResourceFromItem` is a mirror of that

Reply:
Works for me; I was trying to follow the paradigm of  `Marshal<type>`/`Unmarshal<type>`, but `Deitemize` really is a mouthful lol.
How about `ItemFromResource` and `ItemToResource`?  I'd like to preserve the consistency of prefix for the different concrete functions if possible (e.g. `itemizeUser => itemFromUser`, `deitemizeCertAuthority => itemToCertAuthority`, etc).